File: go/addsubs.go
package main

import (
	"fmt"
	"log"
	"os"
	"os/exec"
	"sort"
	"strings"
	"sync"
)

type ProgramError struct {
	Err string
}

const (
	MismatchError string = "Not the same amount of video and sub files"
	LangError            = "Language not supported"
	ExitError            = "User Cancelled"
	InputError           = "IO error"
)

func (pe *ProgramError) Error() string {
	return fmt.Sprintf("Program Error: %v", pe.Err)
}

type Output struct {
	output []byte
	err    error
}

func AddSubs(dir string, videoformat string, subformat string, lang string) ([]Output, error) {
	// ISO 639.2
	langs := map[string]string{
		"jpn": "Japanese",
		"eng": "English",
		"spa": "Spanish",
		"und": "Undetermined",
	}
	if _, ok := langs[lang]; !ok {
		return nil, &ProgramError{LangError}
	}

	// create list of files
	videofiles := make([]string, 0, 12)
	subfiles := make([]string, 0, 12)
	entries, err := os.ReadDir(dir)
	if err != nil {
		return nil, err
	}
	for _, entry := range entries {
		if !entry.IsDir() {
			filename := entry.Name()
			if strings.Contains(filename, videoformat) {
				videofiles = append(videofiles, filename)
			} else if strings.Contains(filename, subformat) {
				subfiles = append(subfiles, filename)
			}
		}
	}

	// Check length
	if len(videofiles) != len(subfiles) {
		return nil, &ProgramError{MismatchError}
	}

	// Show user
	sort.Sort(sort.StringSlice(videofiles))
	sort.Sort(sort.StringSlice(subfiles))
	fmt.Println("Joining sub files to these video files")
	for i, sub := range subfiles {
		fmt.Printf("%s\t%s", sub, videofiles[i])
	}

	// User Confirmation
	fmt.Print("Are these pairs corret? (Y/n): ")
	var confirmation string
	fmt.Scanln(&confirmation)
	if strings.Contains(confirmation, "n") {
		return nil, &ProgramError{ExitError}
	}

	// Make output directory
	cmd := exec.Command("mkdir", dir+"/output")
	err = cmd.Run()
	if err != nil {
		return nil, err
	}

	// Run mkvmerge for all the files
	var wg sync.WaitGroup
	ch := make(chan Output, len(subfiles))
	for i, s := range subfiles {
		wg.Add(1)
		v := videofiles[i]
		defer fmt.Println(v)

		args := strings.Join([]string{
			"-o",
			dir + "/output/" + v,
			v,
			"--language",
			"0:" + lang,
			"--track-name",
			"0:" + langs[lang],
			s}, " ")

		go func(args string) {
			defer wg.Done()
			cmdMerge := exec.Command("mkvmerge", args)
			out, err := cmdMerge.Output()
			ch <- Output{out, err}
		}(args)
	}
	wg.Wait()
	close(ch)
	ret := make([]Output, 0, len(subfiles))
	for output := range ch {
		ret = append(ret, output)
	}

	return ret, nil
}

func main() {
	fmt.Println("vim-go")
	results, err := AddSubs(os.Args[1], os.Args[2], os.Args[3], os.Args[4])
	if err != nil {
		log.Fatal(err)
	}
	for _, output := range results {
		stdout := output.output
		err := output.err
		if err != nil {
			log.Fatal(err)
		}
		fmt.Println(string(stdout))
	}
}

File: c/addsubs.c
#include <asm-generic/errno-base.h>
#include <linux/limits.h>
#define _DEFAULT_SOURCE
#include <dirent.h>
#include <stdbool.h>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <pthread.h>
#include <sys/stat.h> // mkdir
#include <errno.h> // get libc error


enum ProgramError {
	LANGUAGE_ERR = 1,
	DIR_ERR,
	NO_FILE_ERR,
	MISMATCH_ERR,
	RESPONSE_ERR,
	USER_CANCEL,
	OUTPUT_MKDIR_ERR,
	SNPRINTF_ERR,
	SPRINTF_ERR,
	ARG_COUNT_ERR,
	THREAD_CREATE_ERR,
	THREAD_JOIN_ERR,
};

char* program_error_msg(int err) {
	switch (err) {
		case LANGUAGE_ERR:
			return "LANGUAGE_ERR: language not supported.";
		case DIR_ERR:
			return "DIR_ERR: that is not a directory.";
		case NO_FILE_ERR:
			return "NO_FILE_ERR: either there are no video files or sub files that meet the file format specified.";
		case MISMATCH_ERR:
			return "MISMATCH_ERR: no equal ammount of video and sub files.";
		case RESPONSE_ERR:
			return "RESPONSE_ERR: response was too long?";
		case USER_CANCEL:
			return "user cancelled execution.";
		case OUTPUT_MKDIR_ERR:
			return "OUTPUT_MKDIR_ERR: mkdir error out unexepectedly. Check permissios on the directory.";
		case SNPRINTF_ERR:
			return "SNPRINTF_ERR: encoding failed or command line string too long.";
		case SPRINTF_ERR:
			return "SPRINTF_ERR: sprintf encoding failed";
		case ARG_COUNT_ERR:
			return "ARG_COUNT_ERR: wrong amount of arguments.";
		case THREAD_CREATE_ERR:
			return "THREAD_CREATE_ERR: thread create failed.";
		case THREAD_JOIN_ERR:
			return "THREAD_JOIN_ERR: pthread join failed.";
		default:
			return "undefined error";
	}
}

int program_error_puts_return(int err) {
	puts(program_error_msg(err));
	return err;
}


typedef char* string;

#define MAX_PATH 4096



string langs(const string lang) {
	if (strcmp(lang, "jpn") == 0) {
		return "Japanese";
	} else {
		return NULL;
	}
}

int compare_str(const void *a, const void *b) {
	return strcmp((string)a, (string)b);
}

void* run_cmd(void* input) {
	//int ret = system((string)input);
	return NULL;
}



typedef struct {
	char arr[1000][PATH_MAX];
	size_t len;
} array;

array init_arr() {
	array ret = {
		.len = 0,
	};
	return ret;
}


void append_str(array* a, const char* input) {
	strcpy(a->arr[a->len], input);
	a->len += 1;
}

// TODO testing
int addsubs(const string dir, const string videoformat, const string subformat, const string lang) {
	
	string language = langs(lang);
	if (language == NULL) {
		return program_error_puts_return(LANGUAGE_ERR);
	}

	array videofiles = init_arr();
	array subfiles = init_arr();

	DIR *folder = opendir(dir);
	if (folder == NULL) {
		return program_error_puts_return(DIR_ERR);
	}

	struct dirent *entry;
	while ((entry = readdir(folder))) {
		if (entry->d_type == DT_REG) {
			if (strstr(entry->d_name, videoformat) != NULL) {
				append_str(&videofiles, entry->d_name);
			} else if (strstr(entry->d_name, subformat) != NULL) {
				append_str(&videofiles, entry->d_name);
			}
		}
	}
	closedir(folder);

	if (videofiles.len == 0 || subfiles.len == 0) {
		return program_error_puts_return(NO_FILE_ERR);
	}
	if (videofiles.len != subfiles.len) {
		return program_error_puts_return(MISMATCH_ERR);
	}

	qsort(videofiles.arr, videofiles.len, PATH_MAX, compare_str);
	qsort(subfiles.arr, subfiles.len, PATH_MAX, compare_str);

	puts("Are these pairs correct? (Y/n):");
	char response[10];
	if (scanf("%s", response) < 0) {
		return program_error_puts_return(RESPONSE_ERR);
	}
	if (strstr(response, "n") != NULL) {
		return program_error_puts_return(USER_CANCEL);
	}

	const string os = "/output";
	char output_dir[MAX_PATH];
	strcpy(output_dir, dir);
	strcat(output_dir, os);
	int ret = mkdir(output_dir, 0770);
	if (ret != 0 && errno != EEXIST) {
		return program_error_puts_return(OUTPUT_MKDIR_ERR);
	}

	// TODO test loop and join
	pthread_t thread_id;
	for (int i = 0; i < subfiles.len; i++) {
		//string sub = subfiles.ptr[i];
		string vid = videofiles.arr[i];

		const string mkvmerge_cmd_str = "mkvmerge -o %s/%s %s --language 0:%s --track-name 0:%s";
		string cmd = malloc(MAX_PATH*3 + strlen(mkvmerge_cmd_str)  + strlen(lang) + strlen(language) + 3);
		int ret = sprintf(cmd, mkvmerge_cmd_str, output_dir, vid, vid, lang, language);
		if (ret < 0) {
			return program_error_puts_return(SPRINTF_ERR);
		}
		
		ret = pthread_create(&thread_id, NULL, run_cmd, (void*)cmd);
		if (ret != 0) {
			return program_error_puts_return(THREAD_CREATE_ERR);
		}
		free(cmd);	
	}
	ret = pthread_join(thread_id,NULL);
	if (ret != 0) {
		return program_error_puts_return(THREAD_JOIN_ERR);
	}
	return 0;
}



int main(int argc, string argv[]) { 
	if (argc != 5) {
		return program_error_puts_return(ARG_COUNT_ERR);
	}
	int ret = addsubs(argv[1], argv[2], argv[3], argv[4]);
	if (ret != 0) {
		return ret;
	}

	return 0;
}
